codewof/programming/management/commands/_QuestionsLoader.py

```diff
@@ -151,6 +178,74 @@ def load(self):
                 self.mark_translation_availability(question, required_fields=required_fields)
                 question.save()
 
+                # Add programming concepts
+                concept_slugs = question_data.get("concepts", [])
+                for concept_slug in concept_slugs:
+                    text = "not added"
+                    concept = None
+                    if type(concept_slug) is tuple:
+                        text, added_concept = concept_slug
+                        if text != "added":
+                            raise InvalidYAMLValueError(
+                                self.structure_file_path,
+                                "concepts - value '{} {}' - added text is invalid".format(text, slug),
+                            )
+                        concept = added_concept
+                    try:
+                        if text != "added":
+                            concept = ProgrammingConcepts.objects.get(
+                                slug=concept_slug
+                            )
+                        if concept.children.exists() and text == "not added":
+                            # Check if need to add child concept
+                            for child in concept.children:
+                                if child not in concept_slugs:
+                                    concept_slugs.append(("added", child))
+                        # Check if need to add parent concept
+                        if concept.parent is not None and concept.parent not in concept_slugs:
+                            concept_slugs.append(("added", concept.parent))
+                        question.concepts.add(concept)
+                    except ObjectDoesNotExist:
+                        raise KeyNotFoundError(
+                            self.structure_file_path,
+                            concept_slug,
+                            "Concepts"
+                        )
+
+                # Add question contexts
+                context_slugs = question_data.get("contexts", [])
```

Comment:
Another instance of copy-pasted code due to the similarity of Concepts and Contexts.